File: src/shared/DigitalStokvel.Domain/Entities/Payout.cs
using DigitalStokvel.Domain.Common;
using DigitalStokvel.Domain.Enums;

namespace DigitalStokvel.Domain.Entities;

/// <summary>
/// Represents a payout from a group to its members
/// </summary>
public class Payout : BaseEntity
{
    /// <summary>
    /// Reference to the group
    /// </summary>
    public required Guid GroupId { get; set; }

    /// <summary>
    /// Type of payout
    /// </summary>
    public PayoutType PayoutType { get; set; }

    /// <summary>
    /// Total amount to be paid out
    /// </summary>
    public required decimal TotalAmount { get; set; }

    /// <summary>
    /// Current payout status
    /// </summary>
    public PayoutStatus Status { get; set; } = PayoutStatus.PendingApproval;

    /// <summary>
    /// User who initiated the payout
    /// </summary>
    public required Guid InitiatedBy { get; set; }

    /// <summary>
    /// User who approved the payout
    /// </summary>
    public Guid? ApprovedBy { get; set; }

    /// <summary>
    /// Timestamp when payout was initiated
    /// </summary>
    public DateTime InitiatedAt { get; set; }

    /// <summary>
    /// Timestamp when payout was approved
    /// </summary>
    public DateTime? ApprovedAt { get; set; }

    /// <summary>
    /// Timestamp when payout was executed
    /// </summary>
    public DateTime? ExecutedAt { get; set; }

    /// <summary>
    /// Scheduled payout date
    /// </summary>
    public DateTime? ScheduledDate { get; set; }

    /// <summary>
    /// Reason for the payout
    /// </summary>
    public string? Reason { get; set; }

    /// <summary>
    /// Navigation property to Group
    /// </summary>
    public Group? Group { get; set; }

    /// <summary>
    /// Individual disbursements to members
    /// </summary>
    public ICollection<PayoutDisbursement> Disbursements { get; set; } = new List<PayoutDisbursement>();

    /// <summary>
    /// Check if payout requires approval
    /// </summary>
    public bool RequiresApproval => Status == PayoutStatus.PendingApproval;

    /// <summary>
    /// Check if dual approval is complete
    /// </summary>
    public bool IsDualApprovalComplete => InitiatedBy != default && ApprovedBy != default && InitiatedBy != ApprovedBy;
}

File: src/web/DigitalStokvel.WebDashboard/src/types/payout.ts
/**
 * Payout Type Definitions
 * Based on design.md Section 5.4 (Payout API) and Section 4.2 (Database Schema)
 */

export type PayoutType = 'rotating' | 'year_end' | 'emergency';
export type PayoutStatus = 
  | 'pending_chairperson_approval'
  | 'pending_treasurer_approval' 
  | 'approved'
  | 'processing'
  | 'completed'
  | 'failed'
  | 'rejected';
export type DisbursementStatus = 'pending' | 'processing' | 'completed' | 'failed';

// Initiator/Approver Info
export interface PayoutPerson {
  id: string;
  name: string;
  role: 'chairperson' | 'treasurer' | 'secretary' | 'member';
}

// Disbursement (payment to individual member)
export interface PayoutDisbursement {
  id?: string;
  member: {
    id: string;
    name: string;
  };
  amount: number;
  transactionRef?: string;
  status: DisbursementStatus;
  executedAt?: string; // ISO timestamp
  failureReason?: string;
}

// Payout (full payout request)
export interface Payout {
  id: string;
  groupId: string;
  payoutType: PayoutType;
  totalAmount: number;
  status: PayoutStatus;
  initiatedBy: PayoutPerson;
  initiatedAt: string; // ISO timestamp
  approvedBy?: PayoutPerson;
  approvedAt?: string; // ISO timestamp
  executedAt?: string; // ISO timestamp
  requiresApprovalFrom?: 'chairperson' | 'treasurer';
  disbursements?: PayoutDisbursement[];
  note?: string;
  comment?: string; // Approval comment
  executionStatus?: string;
  rejectedBy?: PayoutPerson;
  rejectedAt?: string; // ISO timestamp
  rejectionReason?: string;
}

// Initiate Payout Request
export interface InitiatePayoutRequest {
  groupId: string;
  payoutType: PayoutType;
  recipientMemberId?: string; // Required for rotating, optional for year_end
  amount: number;
  note?: string;
}

// Initiate Payout Response
export interface InitiatePayoutResponse {
  id: string;
  groupId: string;
  payoutType: PayoutType;
  totalAmount: number;
  status: PayoutStatus;
  initiatedBy: PayoutPerson;
  initiatedAt: string;
  requiresApprovalFrom?: string;
}

// Approve Payout Request
export interface ApprovePayoutRequest {
  pin: string; // Encrypted PIN
  comment?: string;
}

// Approve Payout Response
export interface ApprovePayoutResponse {
  id: string;
  status: PayoutStatus;
  approvedBy: PayoutPerson;
  approvedAt: string;
  executionStatus: string;
}

// Reject Payout Request
export interface RejectPayoutRequest {
  pin: string; // Encrypted PIN
  reason: string;
}

// Get Payouts Filters
export interface PayoutFilters {
  groupId?: string;
  status?: PayoutStatus;
  fromDate?: string; // ISO date
  toDate?: string; // ISO date
  page?: number;
  limit?: number;
}

// Pagination
export interface Pagination {
  page: number;
  limit: number;
  total: number;
  pages: number;
}

// List Payouts Response
export interface ListPayoutsResponse {
  payouts: Payout[];
  pagination: Pagination;
}

// Payout Summary (for dashboard)
export interface PayoutSummary {
  totalPayouts: number;
  pendingApprovals: number;
  approvedAwaitingExecution: number;
  completed: number;
  totalAmountDisbursed: number;
}
